kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/bootstrap/FilterChainFactory.java

```diff
@@ -25,39 +26,81 @@
  * required for instantiation at the point at which instances are created.
  * New instances are created during initialization of a downstream channel.
  */
-public class FilterChainFactory {
+public class FilterChainFactory implements AutoCloseable {
 
-    record UninitializedFilterFactory(String instanceName, FilterFactory<? super Object, ? super Object> filterFactory, Object config) {
+    enum FilterFactoryState {
+        UNINITIALIZED,
+        INITIALIZED,
+        CLOSED
 
-        private InitializedFilterFactory initialize(FilterFactoryContext context) {
-            Object result;
+    }
+
+    private static final class Wrapper implements AutoCloseable {
+
+        private final String instanceName;
+        private final FilterFactory<? super Object, ? super Object> filterFactory;
+        private final Object config;
+        private static Object initResult;
+
+        private FilterFactoryState filterFactoryState;
```

Comment:
Potential thread safety issue here, since I think `createFilter` can be called on a different thread than `initialize` and `close`. So we ought to use an `AtomicReference`, unless we can enforce thread safety another way. 

Should probably document the threading guarantees offered by the `FilterChainFactory`, if only for our own benefit.

Reply:
We _could_ drop the check on closedness which I've put in `createFilter()` (and thus not guarantee that there's no race between `FilterFactory.close()` and `FilterFactory.createFilter()`). Coupled with a requirement that the `FilterChainFactory` is created and closed from the same thread this would allow us to do away with the atomic we'd otherwise require. It would be good to try to quantify whether the CAS is really that much of a perf hit though. It's 'only' going to affect pipeline creation, which isn't going to dominate performance in a normal Kafka cluster (where long-lived connections are the norm).

So I'm incluned to go with `AtomicBoolean` here. Wdyt @SamBarker @k-wall ?

Reply:
I'm not sure if it added any actual value but I quite liked the state machine model you had with `FilterFactoryState`. 

That said I don't think we need to agonise over every last CPU cycle in pipeline creation so I'd stick with Atomic* option.